ffi/src/lib.rs

```diff
@@ -1801,7 +1913,9 @@ macro_rules! define_destructor (
     ($name:ident, $t:ty) => (
         #[no_mangle]
         pub unsafe extern "C" fn $name(obj: *mut $t) {
-            let _ = Box::from_raw(obj);
+            if !obj.is_null() {
```

Comment:
I argued against null-checking destructors once before, so I'll do it again.  I contend that if we're double freeing (or freeing null instances!) that something is horribly wrong and a quick crash report is the best possible outcome.  (All the asserts this commit adds should generate exactly those crash reports, hopefully with actionable stacks.)
 
However, both you (@thomcc) and I think @fluffyemily wanted to do this null-checking, and this is definitely not my area of expertise, so can you come to a decision and inform me of what our pattern is?  Thanks!

Reply:
As discussed in IRC, because we can't panic across FFI boundaries doing null checks is probably the best option.